src/ControlSystem/PiecewisePolynomial.cpp

```diff
@@ -0,0 +1,138 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ControlSystem/PiecewisePolynomial.hpp"
+
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/EqualWithinRoundoff.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeArray.hpp"
+
+template <size_t MaxDeriv>
+FunctionOfTime::PiecewisePolynomial<MaxDeriv>::PiecewisePolynomial(
+    double t, value_type initial_func_and_derivs) noexcept
+    : deriv_info_at_update_times_{{t, std::move(initial_func_and_derivs)}} {}
+
+template <size_t MaxDeriv>
+template <size_t MaxDerivReturned>
+std::array<DataVector, MaxDerivReturned + 1>
+FunctionOfTime::PiecewisePolynomial<MaxDeriv>::operator()(const double t) const
+    noexcept {
+  const auto& closest_deriv_info = closest_time_deriv_info(t);
+  const double dt = t - closest_deriv_info.time;
+  const value_type& coefs = closest_deriv_info.derivs_coefs;
+
+  // initialize result for the number of derivs requested
+  std::array<DataVector, MaxDerivReturned + 1> result =
+      make_array<MaxDerivReturned + 1>(DataVector(coefs.back().size(), 0.0));
+
+  // evaluate the polynomial using ddpoly (Numerical Recipes sec 5.1)
+  result[0] = coefs[MaxDeriv];
+  for (size_t j = MaxDeriv; j-- > 0;) {
+    const size_t min_deriv =
+        MaxDerivReturned < (MaxDeriv - j) ? MaxDerivReturned : MaxDeriv - j;
+    for (size_t k = min_deriv; k > 0; k--) {
+      gsl::at(result, k) = gsl::at(result, k) * dt + gsl::at(result, k - 1);
+    }
+    result[0] = result[0] * dt + gsl::at(coefs, j);
+  }
+  // after the first derivative, factorial constants come in
+  double fact = 1.0;
+  for (size_t j = 2; j < MaxDerivReturned + 1; j++) {
+    fact *= j;
```

Comment:
replace with constexpr factorial?

Reply:
Larry, don't be silly, that doesn't help here. But here's an implementation anyway:

```cpp
/*!
 * \ingroup ConstantExpressionsGroup
 * \brief Compute the factorial \f$N!\f$ of type `ReturnType`.
 */
template <size_t N, typename ReturnType>
constexpr ReturnType factorial() noexcept {
  ReturnType fact{1};
  for (size_t j = 2; j < N + 1; j++) {
    fact *= j;
  }
  return fact;
}
```